Step1: Please go to the folder and rt click open terminal

Step2: Please install python forst and add to path python and pip

Step3: pip install pandas

Step4: pip install numpy

Step5: If both are installed then it will show requirement satisfied for both

Step6: To upgrade the installed version use: pip install --upgrade pandas similarly numpy keep in latest version always

Step7: pip install jupyter

Step8: jupyter notebook

A browser will open having information about your notebook created by terminal now select option just after upload and change to python 3 file.

Comment: cntr+/

In [1]:
import pandas as pd
import numpy as np

Now we can import the packages after installing and then to run the cell we need to press: shift+enter
While running it can take time for the first cell but not then
**In[star] means running not excecuted and In[1] means it is executed now. **

Now creatign a dictionary and then creating a dataframe.
A dataframe is like an excel sheet which can help us to analyse huge data easily using numpy feature and process data fast due to faster indexing.

In [2]:
dict1={
    "name":['a','b','c','d'],
    "marks":['11','12','15','14'],
    "city":['p','q','r','s']
}

In [3]:
dict1

{'name': ['a', 'b', 'c', 'd'],
 'marks': ['11', '12', '15', '14'],
 'city': ['p', 'q', 'r', 's']}

In [4]:
df=pd.DataFrame(dict1)

In [5]:
df

name marks city
0    a    11    p
1    b    12    q
2    c    15    r
3    d    14    s

As we know as the dataframe is created using a function as:
df=pd.DataFrame(dict1)
And we can see data in table form and we can see that there are numbers as 0 to 3 which are called as INDEX and columns have name.
The number given to IN[] is defined as the number of which it is being executed this helps for better understanding.

Now we need to import data of datafrome in excel or csv as:
df.to_csv('name_file.csv')
We can see int he home section tab we can see the file there as home is our fiolder whic we created for pandas 
We can see in the folder too it is present.
Data EVEN with indox and column will be present here.

Index is not needed then:
df.to_csv('first_index_false.csv',index=False)

In [6]:
df.to_csv('first.csv')

In [7]:
df.to_csv('first_index_false.csv',index=False)

In case of huge dataframe It cannot be seen complete only top 5 rows can be seen.
If we need to see a specific row we need to use code:
df.head(2)
This is showing first 2 rows
df.tail(2)
This is showing last 2 rows
raveesh= pd.read_csv('Raveesh1.csv')
This command is used for reading an excel file from the system

In [8]:
df.head(2)

name marks city
0    a    11    p
1    b    12    q

In [9]:
df.tail(2)

name marks city
2    c    15    r
3    d    14    s

In [10]:
df.describe()

name marks city
count     4     4    4
unique    4     4    4
top       a    11    p
freq      1     1    1

In [11]:
raveesh= pd.read_csv('Raveesh1.csv')

In [12]:
raveesh

Unnamed: 0.19  Unnamed: 0.18  Unnamed: 0.17  Unnamed: 0.16  Unnamed: 0.15  \
0               0              0              0              0              0   
1               1              1              1              1              1   
2               2              2              2              2              2   
3               3              3              3              3              3   
4               4              4              4              4              4   
5               5              5              5              5              5   
6               6              6              6              6              6   
7               7              7              7              7              7   
8               8              8              8              8              8   
9               9              9              9              9              9   
10             10             10             10             10             10   
11             11             11             11             11             11   
12             12             12             12             12             12   
13             13             13             13             13             13   
14             14             14             14             14             14   
15             15             15             15             15             15   
16             16             16             16             16             16   
17             17             17             17             17             17   
18             18             18             18             18             18   
19             19             19             19             19             19   
20             20             20             20             20             20   
21             21             21             21             21             21   
22             22             22             22             22             22   
23             23             23             23             23             23   
24             24             24             24             24             24   
25             25             25             25             25             25   
26             26             26             26             26             26   
27             27             27             27             27             27   
28             28             28             28             28             28   
29             29             29             29             29             29   
30             30             30             30             30             30   
31             31             31             31             31             31   
32             32             32             32             32             32   
33             33             33             33             33             33   
34             34             34             34             34             34   
35             35             35             35             35             35   
36             36             36             36             36             36   
37             37             37             37             37             37   
38             38             38             38             38             38   
39             39             39             39             39             39   
40             40             40             40             40             40   
41             41             41             41             41             41   
42             42             42             42             42             42   
43             43             43             43             43             43   
44             44             44             44             44             44   
45             45             45             45             45             45   
46             46             46             46             46             46   

    Unnamed: 0.14  Unnamed: 0.13  Unnamed: 0.12  Unnamed: 0.11  Unnamed: 0.10  \
0               0              0  

We need to change the data or bring a definite data from the above data frame:

Data from the region will be provided using:

raveesh['Region'][0]

This shows the value of column region in cell 0 or index 0

and changed using:

raveesh.at[0,'Region']="West"

raveesh['Region'][0]='Central'

raveesh['Units'][0]=4.5678

BOTH GOT CHANGED WITH THE GIVEN WARNING

C:\Users\versha pande\AppData\Local\Temp\ipykernel_4168\3685774197.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

raveesh['Units'][0]=4.5678

CHANGING THE INDEX IN THE TABLE:

r1.index=['x','y','z','w']


In [13]:
raveesh['Region'][0]

'Central'

In [14]:
raveesh.at[0,'Region']="West"

In [15]:
raveesh['Region'][0]

'West'

In [16]:
raveesh['Units'][0]=4.5678

C:\Users\versha pande\AppData\Local\Temp\ipykernel_19488\3685774197.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  raveesh['Units'][0]=4.5678


In [17]:
raveesh['Units'][0]

4.5678

In [18]:
raveesh['Region'][0]='Central'

C:\Users\versha pande\AppData\Local\Temp\ipykernel_19488\1276026482.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  raveesh['Region'][0]='Central'


In [19]:
raveesh['Region'][0]

'Central'

Make is save in the existing file we need to save the changed data then:

raveesh.to_csv('Raveesh1.csv')

Data got changed here.

Change the index here from 0 1 2 3 to w x y z:


In [20]:
raveesh.to_csv('Raveesh1.csv')

In [21]:
r1=pd.read_csv('first_index_false.csv')


In [22]:
r1


name  marks city
0    a     11    p
1    b     12    q
2    c     15    r
3    d     14    s

In [23]:
r1.index=['x','y','z','w']


In [24]:
r1

name  marks city
x    a     11    p
y    b     12    q
z    c     15    r
w    d     14    s

WHY PANDAS:

1 It is an open source library coded in python which provides better tools for data analysis. In place of excel it is used because here we can analyse data by using python code and can use multiple functions provided by pythion which may not be available with excel or very difficult to impliment. And without pandas only python using dictionary list tupples etc we cannot analyse more.

2 It uses the leveragw of numpy to increase its speed for better calculations and also provide good support as we know that numpy library is coded in C LANGUAGE which is properly used by pandas for data analysis.

3 It provides high and robust data operations: change the value append the values adding or removing the values in csv files etc.



DATA STRUCTURES IN PANDAS:

Mainly of 2 types: Series and DataFrame

Series: A 1 dimentional array having same datatype it is in a dataframe which is a specific column like name column all string.

Data Frame: A tabular spread sheet like struction having muktiple rows which contains 1 or more then 1 columns.

A 2D array having different datatypes but columns or series having same datatype.



In [25]:
type(raveesh)

pandas.core.frame.DataFrame

In [26]:
type(raveesh['Region'])

pandas.core.series.Series

In [27]:
# Printing the series of random numbers using python and numpy.

ser=pd.Series(np.random.rand(34))

# 34 is the number of rows. 

In [28]:
ser

0     0.460776
1     0.364559
2     0.368126
3     0.871356
4     0.517078
5     0.158805
6     0.061424
7     0.425506
8     0.068125
9     0.023854
10    0.440941
11    0.352194
12    0.063481
13    0.495738
14    0.147753
15    0.448119
16    0.734702
17    0.090984
18    0.695122
19    0.139847
20    0.520140
21    0.454380
22    0.738554
23    0.719760
24    0.089246
25    0.136749
26    0.736452
27    0.378660
28    0.533892
29    0.977889
30    0.865889
31    0.829367
32    0.946217
33    0.356879
dtype: float64

In [29]:
newdf=pd.DataFrame(np.random.rand(334,5))
# Here creating a new dataframe with rows as 334 and columns as 5

In [30]:
newdf

0         1         2         3         4
0    0.073378  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2    0.754490  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [31]:
newdf.describe()
# All basic data analytics on the data can be seen 

0           1           2           3           4
count  334.000000  334.000000  334.000000  334.000000  334.000000
mean     0.478494    0.519182    0.493359    0.493181    0.498091
std      0.292952    0.285230    0.278990    0.290860    0.290970
min      0.000425    0.002367    0.010260    0.001984    0.007259
25%      0.209985    0.275466    0.261611    0.236893    0.265444
50%      0.459287    0.512726    0.489811    0.501786    0.491215
75%      0.720346    0.771691    0.710637    0.745956    0.748875
max      0.996134    0.998658    0.998451    0.996894    0.999796

In [32]:
newdf.dtypes
# All the datatype of the columns are given 
# By default the datatype for all the columns is same. 
# If we change the value of the first column and first row that is first cell then the column dtype will be changed to object  

0    float64
1    float64
2    float64
3    float64
4    float64
dtype: object

In [33]:
newdf[0][0]="Raveesh"

In [34]:
newdf.dtypes

0     object
1    float64
2    float64
3    float64
4    float64
dtype: object

In [35]:
newdf.head()

# Shows top 5 rows of the dataframe. 

0         1         2         3         4
0   Raveesh  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [36]:
# Printing the index: 
# The output is a rangeindex object 

newdf.index

RangeIndex(start=0, stop=334, step=1)

In [37]:
# Printing the columns: 

newdf.columns

RangeIndex(start=0, stop=5, step=1)

In [38]:
# Convert the dataframe into a numpy array then: 
newdf.to_numpy()
# dtypeis object as we have raveeshin column noew chagning it to a number again: 

array([['Raveesh', 0.43147051609744624, 0.8047946227651648,
        0.7871341188065668, 0.5856449394237557],
       [0.9520116687094761, 0.4185812847780783, 0.3402533332620439,
        0.05613948695158133, 0.9221344716477936],
       [0.7544903132063555, 0.8962170438602716, 0.8452241017622163,
        0.3971829279275445, 0.6549363112853314],
       ...,
       [0.6762182627181952, 0.003854029226648903, 0.43437516109803875,
        0.7251339437218641, 0.5462912779881681],
       [0.07153458870807683, 0.2813925581351864, 0.11275301148850247,
        0.9291002680876805, 0.10299304701585177],
       [0.19850197139332326, 0.012161962672330051, 0.3738543698296205,
        0.47308462957327724, 0.5018515674714138]], dtype=object)

In [39]:
newdf[0][0]=0.345

In [40]:
newdf.head()

0         1         2         3         4
0     0.345  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [41]:
newdf.to_numpy()
# Still Dtype is an object type only 

array([[0.345, 0.43147051609744624, 0.8047946227651648,
        0.7871341188065668, 0.5856449394237557],
       [0.9520116687094761, 0.4185812847780783, 0.3402533332620439,
        0.05613948695158133, 0.9221344716477936],
       [0.7544903132063555, 0.8962170438602716, 0.8452241017622163,
        0.3971829279275445, 0.6549363112853314],
       ...,
       [0.6762182627181952, 0.003854029226648903, 0.43437516109803875,
        0.7251339437218641, 0.5462912779881681],
       [0.07153458870807683, 0.2813925581351864, 0.11275301148850247,
        0.9291002680876805, 0.10299304701585177],
       [0.19850197139332326, 0.012161962672330051, 0.3738543698296205,
        0.47308462957327724, 0.5018515674714138]], dtype=object)

In [42]:
# Transpose of a numoy array: 
newdf.T
# Now number of columns are number of rows and number of rows are number of columns. 

0         1         2         3         4         5         6    \
0     0.345  0.952012   0.75449  0.338647  0.864826  0.748116  0.098416   
1  0.431471  0.418581  0.896217  0.738839  0.155727  0.293385  0.104519   
2  0.804795  0.340253  0.845224  0.414045  0.812999  0.419558  0.434464   
3  0.787134  0.056139  0.397183  0.248324  0.481722  0.487839  0.095072   
4  0.585645  0.922134  0.654936  0.916931  0.518476  0.374039  0.783167   

        7         8         9    ...       324       325       326       327  \
0   0.08425   0.17599  0.040376  ...  0.983477  0.641108  0.570025  0.684112   
1   0.83102  0.009273  0.935321  ...  0.815418  0.750228  0.965918  0.535168   
2  0.550588  0.946513  0.093772  ...  0.212573  0.877955  0.484359  0.200098   
3  0.198276  0.614534  0.267937  ...   0.97229   0.69143  0.176613   0.58141   
4  0.916422  0.071777  0.967133  ...  0.023628  0.714042  0.370831  0.501366   

        328       329       330       331       332       333  
0   0.20549  0.876511  0.824137  0.676218  0.071535  0.198502  
1  0.313919   0.92141  0.797805  0.003854  0.281393  0.012162  
2  0.852029  0.620888  0.903213  0.434375  0.112753  0.373854  
3   0.98564  0.802798  0.263599  0.725134    0.9291  0.473085  
4  0.689911  0.665418  0.628052  0.546291  0.102993  0.501852  

[5 rows x 334 columns]

In [43]:
# SORTING THE NUMPY ARRAY: 

# In terms of rows: We need to put axis=0 and ascending='False' which is by deafult true

# newdf.sort_index(axis=0,ascending=False)

# In terms of columns: We need to put axis=1 and ascending='False' which is by deafult true

newdf.head()


0         1         2         3         4
0     0.345  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [44]:
newdf.sort_index(axis=0,ascending=False)

0         1         2         3         4
333  0.198502  0.012162  0.373854  0.473085  0.501852
332  0.071535  0.281393  0.112753  0.929100  0.102993
331  0.676218  0.003854  0.434375  0.725134  0.546291
330  0.824137  0.797805  0.903213  0.263599  0.628052
329  0.876511  0.921410  0.620888  0.802798  0.665418
..        ...       ...       ...       ...       ...
4    0.864826  0.155727  0.812999  0.481722  0.518476
3    0.338647  0.738839  0.414045  0.248324  0.916931
2     0.75449  0.896217  0.845224  0.397183  0.654936
1    0.952012  0.418581  0.340253  0.056139  0.922134
0       0.345  0.431471  0.804795  0.787134  0.585645

[334 rows x 5 columns]

In [45]:
newdf.sort_index(axis=1,ascending=False)

4         3         2         1         0
0    0.585645  0.787134  0.804795  0.431471     0.345
1    0.922134  0.056139  0.340253  0.418581  0.952012
2    0.654936  0.397183  0.845224  0.896217   0.75449
3    0.916931  0.248324  0.414045  0.738839  0.338647
4    0.518476  0.481722  0.812999  0.155727  0.864826
..        ...       ...       ...       ...       ...
329  0.665418  0.802798  0.620888  0.921410  0.876511
330  0.628052  0.263599  0.903213  0.797805  0.824137
331  0.546291  0.725134  0.434375  0.003854  0.676218
332  0.102993  0.929100  0.112753  0.281393  0.071535
333  0.501852  0.473085  0.373854  0.012162  0.198502

[334 rows x 5 columns]

In [46]:
# Type of 1 column if the dataframe 

newdf.head()

0         1         2         3         4
0     0.345  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [47]:
newdf[0]

0         0.345
1      0.952012
2       0.75449
3      0.338647
4      0.864826
         ...   
329    0.876511
330    0.824137
331    0.676218
332    0.071535
333    0.198502
Name: 0, Length: 334, dtype: object

In [48]:
type(newdf[0])
# This shows combination of series makes a dataframe. 

pandas.core.series.Series

In [49]:
# Difference between view and copy: 

newdf2=newdf

# Now what will happen will newdf2 a new dataframe will be created. 
# No newdf2 will just be a view of newdf 

Say we change the value of 1 cell in newdf2 then same will be implimented in newdf also as newdf2 is a view of newdf it is not a different dataframe

In [50]:
newdf2[0][0]=0.5678

In [51]:
newdf.head()

0         1         2         3         4
0    0.5678  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

Here we can see that we have changed the newdf2 but the changes are seen in newdf dataframe because newdf2 is not a copy it is just a view of same daatframe which newdf is also having. 

View is a similar concept in database like the dataframe is at 1 place in memory which is being pointed by 2 variables newdf and newdf2.

In [52]:
# To copy we can use the below command. 

newdf2=newdf.copy()

# copy function have an argument deep which is bydefault True that means a copy is created or else if False then a view is created. 

# newdf3 is creating a view and newdf2 is creating a copy that means now there are 2 different dataframe in the memory newdf and newdf3 as 1 and newdf2 as other one.  

In [53]:
newdf3=newdf[:]

In [54]:
newdf3[0][0]=0.234

C:\Users\versha pande\AppData\Local\Temp\ipykernel_19488\4068841478.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdf3[0][0]=0.234


In [55]:
newdf3.head()

0         1         2         3         4
0     0.234  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [56]:
newdf.head()

0         1         2         3         4
0     0.234  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [57]:
newdf2[0][0]=0.2344

C:\Users\versha pande\AppData\Local\Temp\ipykernel_19488\57273243.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdf2[0][0]=0.2344


In [58]:
newdf2.head()

0         1         2         3         4
0    0.2344  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [59]:
newdf.head()

0         1         2         3         4
0     0.234  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

WARNING IN THE FUNCTION:

*C:\Users\versha pande\AppData\Local\Temp\ipykernel_45268\57273243.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame
See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdf2[0][0]=0.2344*
  
This warning is comming because in pandas there are 2 types a dataframe can be used that is voiew and copy bu sometimes pandas dont know what to return a copy or a view hencethee is a chance that in place of copy changes are done in view format which will change the original dataframe **hence we should not change the vaue by using:
newdf3[0][0]=0.234**

Instead we need to use **.loc[]** function:
**newdf2.loc[0,0]=0.7685**

In [60]:
newdf2.loc[0,0]=0.7685

In [61]:
newdf.head()

0         1         2         3         4
0     0.234  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [62]:
newdf2.head()

0         1         2         3         4
0    0.7685  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [63]:
newdf3.loc[0,0]=0.5467

In [64]:
newdf.head()

0         1         2         3         4
0    0.5467  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [65]:
newdf3.head()

0         1         2         3         4
0    0.5467  0.431471  0.804795  0.787134  0.585645
1  0.952012  0.418581  0.340253  0.056139  0.922134
2   0.75449  0.896217  0.845224  0.397183  0.654936
3  0.338647  0.738839  0.414045  0.248324  0.916931
4  0.864826  0.155727  0.812999  0.481722  0.518476

In [66]:
# Changing the name of the columns and index: 

df

# way 1: df.columns=['name','marks','city']
# way 2: newdf3.columns=list("ABCDE") This is only used we we are using some series thing not used for different name.

name marks city
0    a    11    p
1    b    12    q
2    c    15    r
3    d    14    s

In [67]:
df.columns=['name','marks','city']

In [68]:
df

name marks city
0    a    11    p
1    b    12    q
2    c    15    r
3    d    14    s

In [69]:
newdf3

0         1         2         3         4
0      0.5467  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [70]:
newdf3.columns=list("ABCDE")

In [71]:
newdf3

A         B         C         D         E
0      0.5467  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

Now as we know that the column name is channged hence we need to change the code also:

Inicially: **newdf3.loc[0,0]=0.5467** Here we use this so it will create a new column name 0 as in the newdf3 now we dont have 0 COLUMN and in the 0th row we will get a value rest all are NaN.

Now: **newdf3.loc[0,'A']=0.54678** Here we have used the column name hence a new column will NOT be created.

As we know a new 0 column is created hence it is needed to be removed for that we need to use the drop command:

**newdf3.drop(0,axis=1)** This makes column to be dropped where axis=0 means row will be dropped.

In [72]:
newdf3.loc[0,0]=0.5467

# Herwe we will see a creating of new column 0 

C:\Users\versha pande\AppData\Local\Temp\ipykernel_19488\1449317198.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdf3.loc[0,0]=0.5467


In [73]:
newdf3

A         B         C         D         E       0
0      0.5467  0.431471  0.804795  0.787134  0.585645  0.5467
1    0.952012  0.418581  0.340253  0.056139  0.922134     NaN
2     0.75449  0.896217  0.845224  0.397183  0.654936     NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931     NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476     NaN
..        ...       ...       ...       ...       ...     ...
329  0.876511  0.921410  0.620888  0.802798  0.665418     NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052     NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291     NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993     NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852     NaN

[334 rows x 6 columns]

In [74]:
# Now changing the code

newdf3.loc[0,'A']=0.54678

In [75]:
newdf3
# We can see the change int he correct table column.

A         B         C         D         E       0
0     0.54678  0.431471  0.804795  0.787134  0.585645  0.5467
1    0.952012  0.418581  0.340253  0.056139  0.922134     NaN
2     0.75449  0.896217  0.845224  0.397183  0.654936     NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931     NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476     NaN
..        ...       ...       ...       ...       ...     ...
329  0.876511  0.921410  0.620888  0.802798  0.665418     NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052     NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291     NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993     NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852     NaN

[334 rows x 6 columns]

In [76]:
newdf3.drop(0,axis=1)
# 0 column name is now removed.
# axis is needed to be defined bydefualt it is 0 that is row will be removed. 

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [77]:
# newdf.drop('A',axis=1)
# Columns A added to newdf accidently is now removed
# Now it is needed to get selected columns and selected rows from the datset 
# Here we are just returning the view of the dataframe no new dataframe is crerated. 
newdf3.loc[[1,2],['C','D']]

C         D
1  0.340253  0.056139
2  0.845224  0.397183

In [78]:
# Similarly returning all the rows for selected column:
newdf3.loc[:,['C','D']]

C         D
0    0.804795  0.787134
1    0.340253  0.056139
2    0.845224  0.397183
3    0.414045  0.248324
4    0.812999  0.481722
..        ...       ...
329  0.620888  0.802798
330  0.903213  0.263599
331  0.434375  0.725134
332  0.112753  0.929100
333  0.373854  0.473085

[334 rows x 2 columns]

In [79]:
# Similarly returning all the columns for selected rows:
newdf3.loc[[1,2],:]

A         B         C         D         E   0
1  0.952012  0.418581  0.340253  0.056139  0.922134 NaN
2   0.75449  0.896217  0.845224  0.397183  0.654936 NaN

In [80]:
newdf3.drop(0,axis=1)

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [81]:
# Finding the data depending on the query or condition. 

newdf3.loc[(newdf3['A'])<0.3]

A         B         C         D         E   0
6    0.098416  0.104519  0.434464  0.095072  0.783167 NaN
7     0.08425  0.831020  0.550588  0.198276  0.916422 NaN
8     0.17599  0.009273  0.946513  0.614534  0.071777 NaN
9    0.040376  0.935321  0.093772  0.267937  0.967133 NaN
16   0.185661  0.047866  0.050055  0.309682  0.986242 NaN
..        ...       ...       ...       ...       ...  ..
318  0.118219  0.204692  0.617872  0.835479  0.687943 NaN
320  0.168821  0.521379  0.583059  0.032431  0.130951 NaN
328   0.20549  0.313919  0.852029  0.985640  0.689911 NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993 NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852 NaN

[109 rows x 6 columns]

In [82]:
# Using multiple condition. 

newdf3.loc[(newdf3['A']<0.3) & (newdf3['C']>0.1)]

A         B         C         D         E   0
6    0.098416  0.104519  0.434464  0.095072  0.783167 NaN
7     0.08425  0.831020  0.550588  0.198276  0.916422 NaN
8     0.17599  0.009273  0.946513  0.614534  0.071777 NaN
18   0.213053  0.318510  0.877613  0.675313  0.193116 NaN
20   0.288702  0.819112  0.598081  0.778118  0.109224 NaN
..        ...       ...       ...       ...       ...  ..
318  0.118219  0.204692  0.617872  0.835479  0.687943 NaN
320  0.168821  0.521379  0.583059  0.032431  0.130951 NaN
328   0.20549  0.313919  0.852029  0.985640  0.689911 NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993 NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852 NaN

[98 rows x 6 columns]

In [83]:
# Using multiple condition. 

newdf3.loc[((newdf3['A']<0.3) & (newdf3['C']>0.1))|(newdf3['D']>0.1)]

A         B         C         D         E       0
0     0.54678  0.431471  0.804795  0.787134  0.585645  0.5467
2     0.75449  0.896217  0.845224  0.397183  0.654936     NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931     NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476     NaN
5    0.748116  0.293385  0.419558  0.487839  0.374039     NaN
..        ...       ...       ...       ...       ...     ...
329  0.876511  0.921410  0.620888  0.802798  0.665418     NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052     NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291     NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993     NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852     NaN

[307 rows x 6 columns]

Difference between: **loc[] and iloc[]**

**loc[]: This is used when we have column or row names in the argument hence can be used for specific case only if the number of column and rows are huge then we cannot name them all or write the names in the code.**

**iloc[]: This is used when we need to provide index which doesnt matter which column name we have the code iterates with index hence it is benificial for large chunk of data.**

**As per rule of indexing both row and column in iloc starts from 0.**


In [84]:
# previously in loc[]

newdf3.loc[0,'D']

# This gives the cell value of 1st row and the D column or the 4th column. 

0.7871341188065668

In [85]:
# In case of iloc

newdf3.iloc[0,3]

# Here we are getting the same value where ilic[0,3] where 0 is the row and 3 is the column index where row stars from 0 and column start from 0 too.


0.7871341188065668

In [86]:
# Finding selective data using iloc:
newdf3.iloc[[0,1],[1,2]]
# This means that we need row number 0 and 1 and column number 2 and 3 that is B and C


B         C
0  0.431471  0.804795
1  0.418581  0.340253

In [87]:
newdf3.iloc[[0,5,3],[1,2]]
# Here row needed is 0,5,3 only

B         C
0  0.431471  0.804795
5  0.293385  0.419558
3  0.738839  0.414045

In [88]:
# Dropping a row and column:

newdf3.drop(0)

# Bydefault axis is equal to 0 hence row will be taken. 

A         B         C         D         E   0
1    0.952012  0.418581  0.340253  0.056139  0.922134 NaN
2     0.75449  0.896217  0.845224  0.397183  0.654936 NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931 NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476 NaN
5    0.748116  0.293385  0.419558  0.487839  0.374039 NaN
..        ...       ...       ...       ...       ...  ..
329  0.876511  0.921410  0.620888  0.802798  0.665418 NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052 NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291 NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993 NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852 NaN

[333 rows x 6 columns]

In [89]:
newdf3
# But here we can see 0 because here by dropping we are just creating a view my ORIGINAL DF IS NOT CHANGED.

A         B         C         D         E       0
0     0.54678  0.431471  0.804795  0.787134  0.585645  0.5467
1    0.952012  0.418581  0.340253  0.056139  0.922134     NaN
2     0.75449  0.896217  0.845224  0.397183  0.654936     NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931     NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476     NaN
..        ...       ...       ...       ...       ...     ...
329  0.876511  0.921410  0.620888  0.802798  0.665418     NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052     NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291     NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993     NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852     NaN

[334 rows x 6 columns]

In [90]:
# Similarly for columns:
newdf3.drop(0,axis=1)

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [91]:
newdf3
# Same it is still there 0th column. 

A         B         C         D         E       0
0     0.54678  0.431471  0.804795  0.787134  0.585645  0.5467
1    0.952012  0.418581  0.340253  0.056139  0.922134     NaN
2     0.75449  0.896217  0.845224  0.397183  0.654936     NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931     NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476     NaN
..        ...       ...       ...       ...       ...     ...
329  0.876511  0.921410  0.620888  0.802798  0.665418     NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052     NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291     NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993     NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852     NaN

[334 rows x 6 columns]

In [92]:
# Now to permanently delete we need to:
# As newdf3 is not a view but a copy of newdf hence only newdf3 wiil be changed. 
newdf3=newdf3.drop(0,axis=1)

In [93]:
newdf3
# COLUMN 0 is removed.

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [94]:
newdf
# This is not changed. 

0         1         2         3         4
0     0.54678  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [95]:
# Dropping muktiple columns in a dataframe:
newdf3.drop(['A','D'],axis=1)

B         C         E
0    0.431471  0.804795  0.585645
1    0.418581  0.340253  0.922134
2    0.896217  0.845224  0.654936
3    0.738839  0.414045  0.916931
4    0.155727  0.812999  0.518476
..        ...       ...       ...
329  0.921410  0.620888  0.665418
330  0.797805  0.903213  0.628052
331  0.003854  0.434375  0.546291
332  0.281393  0.112753  0.102993
333  0.012162  0.373854  0.501852

[334 rows x 3 columns]

In [96]:
# Here also it is not permanently removed it is just removed in view nothing is harmed in newdf3
# To permanently change it either we need to euate it to newdf3 or we need to use an attribute inplace
# Adding a new column in data frame:
newdf3.loc[0,0]=0.4567

In [97]:
newdf3

A         B         C         D         E       0
0     0.54678  0.431471  0.804795  0.787134  0.585645  0.4567
1    0.952012  0.418581  0.340253  0.056139  0.922134     NaN
2     0.75449  0.896217  0.845224  0.397183  0.654936     NaN
3    0.338647  0.738839  0.414045  0.248324  0.916931     NaN
4    0.864826  0.155727  0.812999  0.481722  0.518476     NaN
..        ...       ...       ...       ...       ...     ...
329  0.876511  0.921410  0.620888  0.802798  0.665418     NaN
330  0.824137  0.797805  0.903213  0.263599  0.628052     NaN
331  0.676218  0.003854  0.434375  0.725134  0.546291     NaN
332  0.071535  0.281393  0.112753  0.929100  0.102993     NaN
333  0.198502  0.012162  0.373854  0.473085  0.501852     NaN

[334 rows x 6 columns]

In [98]:
newdf3.drop(0,axis=1,inplace=True)
# Dropped permanently.

In [99]:
newdf3

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1    0.952012  0.418581  0.340253  0.056139  0.922134
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[334 rows x 5 columns]

In [100]:
# Say we need to drop some rows then:
newdf3.drop([1,5],axis=0,inplace=True)

In [101]:
newdf3
# Here we can see that row number 1 and 5 are dropped but the index is also mismatched now.

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
2     0.75449  0.896217  0.845224  0.397183  0.654936
3    0.338647  0.738839  0.414045  0.248324  0.916931
4    0.864826  0.155727  0.812999  0.481722  0.518476
6    0.098416  0.104519  0.434464  0.095072  0.783167
..        ...       ...       ...       ...       ...
329  0.876511  0.921410  0.620888  0.802798  0.665418
330  0.824137  0.797805  0.903213  0.263599  0.628052
331  0.676218  0.003854  0.434375  0.725134  0.546291
332  0.071535  0.281393  0.112753  0.929100  0.102993
333  0.198502  0.012162  0.373854  0.473085  0.501852

[332 rows x 5 columns]

In [102]:
# To change the index we need to use a function as:
newdf3.reset_index()
# Index is changed but a new column index is created which is not needed hence we can drop it. And to save we need to use inplace

index         A         B         C         D         E
0        0   0.54678  0.431471  0.804795  0.787134  0.585645
1        2   0.75449  0.896217  0.845224  0.397183  0.654936
2        3  0.338647  0.738839  0.414045  0.248324  0.916931
3        4  0.864826  0.155727  0.812999  0.481722  0.518476
4        6  0.098416  0.104519  0.434464  0.095072  0.783167
..     ...       ...       ...       ...       ...       ...
327    329  0.876511  0.921410  0.620888  0.802798  0.665418
328    330  0.824137  0.797805  0.903213  0.263599  0.628052
329    331  0.676218  0.003854  0.434375  0.725134  0.546291
330    332  0.071535  0.281393  0.112753  0.929100  0.102993
331    333  0.198502  0.012162  0.373854  0.473085  0.501852

[332 rows x 6 columns]

In [103]:
newdf3.reset_index(drop=True)
# Confirmation that drop is working fine only index column is dropped.

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1     0.75449  0.896217  0.845224  0.397183  0.654936
2    0.338647  0.738839  0.414045  0.248324  0.916931
3    0.864826  0.155727  0.812999  0.481722  0.518476
4    0.098416  0.104519  0.434464  0.095072  0.783167
..        ...       ...       ...       ...       ...
327  0.876511  0.921410  0.620888  0.802798  0.665418
328  0.824137  0.797805  0.903213  0.263599  0.628052
329  0.676218  0.003854  0.434375  0.725134  0.546291
330  0.071535  0.281393  0.112753  0.929100  0.102993
331  0.198502  0.012162  0.373854  0.473085  0.501852

[332 rows x 5 columns]

In [104]:
# Saving the changes as:
newdf3.reset_index(drop=True,inplace=True)

In [105]:
newdf3
# Saved and index is also properlty maintained.

A         B         C         D         E
0     0.54678  0.431471  0.804795  0.787134  0.585645
1     0.75449  0.896217  0.845224  0.397183  0.654936
2    0.338647  0.738839  0.414045  0.248324  0.916931
3    0.864826  0.155727  0.812999  0.481722  0.518476
4    0.098416  0.104519  0.434464  0.095072  0.783167
..        ...       ...       ...       ...       ...
327  0.876511  0.921410  0.620888  0.802798  0.665418
328  0.824137  0.797805  0.903213  0.263599  0.628052
329  0.676218  0.003854  0.434375  0.725134  0.546291
330  0.071535  0.281393  0.112753  0.929100  0.102993
331  0.198502  0.012162  0.373854  0.473085  0.501852

[332 rows x 5 columns]

THis is the end of **BASIC DATA FRAME** which was needed rest we need to learn by searching only.

Usage of **.loc[] or .iloc[]** Because if we try to change it directly we can get some error or maybe it wont change because we dont know that if our dataframe is a view or a copy hence we need to do this using .loc or .iloc everytime so that surely the value will be changed.

In [106]:
# Using isnull function:
# This function will return a boolean value where if we have some data value returned will be false or else will be true
newdf3['A'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
327    False
328    False
329    False
330    False
331    False
Name: A, Length: 332, dtype: bool

In [107]:
# Say we need to change the value of a column to none then:
newdf3.loc[:,'A']=None
# All rows in column A.

In [108]:
newdf3['A'].isnull()

0      True
1      True
2      True
3      True
4      True
       ... 
327    True
328    True
329    True
330    True
331    True
Name: A, Length: 332, dtype: bool

In [109]:
newdf3
# This is saved now.

A         B         C         D         E
0    None  0.431471  0.804795  0.787134  0.585645
1    None  0.896217  0.845224  0.397183  0.654936
2    None  0.738839  0.414045  0.248324  0.916931
3    None  0.155727  0.812999  0.481722  0.518476
4    None  0.104519  0.434464  0.095072  0.783167
..    ...       ...       ...       ...       ...
327  None  0.921410  0.620888  0.802798  0.665418
328  None  0.797805  0.903213  0.263599  0.628052
329  None  0.003854  0.434375  0.725134  0.546291
330  None  0.281393  0.112753  0.929100  0.102993
331  None  0.012162  0.373854  0.473085  0.501852

[332 rows x 5 columns]

In [110]:
newdf3.loc[:,'A']=32

C:\Users\versha pande\AppData\Local\Temp\ipykernel_19488\3842081037.py:1: FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  newdf3.loc[:,'A']=32


In [111]:
newdf3

A         B         C         D         E
0    32  0.431471  0.804795  0.787134  0.585645
1    32  0.896217  0.845224  0.397183  0.654936
2    32  0.738839  0.414045  0.248324  0.916931
3    32  0.155727  0.812999  0.481722  0.518476
4    32  0.104519  0.434464  0.095072  0.783167
..   ..       ...       ...       ...       ...
327  32  0.921410  0.620888  0.802798  0.665418
328  32  0.797805  0.903213  0.263599  0.628052
329  32  0.003854  0.434375  0.725134  0.546291
330  32  0.281393  0.112753  0.929100  0.102993
331  32  0.012162  0.373854  0.473085  0.501852

[332 rows x 5 columns]

SOME DIFFERENT FUNCTIONS:

In [112]:
df1=pd.DataFrame({"name":['Ab','cd','df'],
                    "toy":['xy','zs',np.nan],
                    "born":[pd.NaT,pd.Timestamp("2021-04-23"),pd.NaT]})

In [113]:
df1

name  toy       born
0   Ab   xy        NaT
1   cd   zs 2021-04-23
2   df  NaN        NaT

Usage of function like :

**dropna()**: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html

**drop_duplicates()**:https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html


In [114]:
# Basic commands:
df1.shape
# Fisrt index is number of rows followed by number of columns.

(3, 3)

In [115]:
df1

name  toy       born
0   Ab   xy        NaT
1   cd   zs 2021-04-23
2   df  NaN        NaT

In [116]:
df1.info()
# info method used for general information about the dataframe.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   name    3 non-null      object        
 1   toy     2 non-null      object        
 2   born    1 non-null      datetime64[ns]
dtypes: datetime64[ns](1), object(2)
memory usage: 200.0+ bytes


In [117]:
# value counts:
df1['toy'].value_counts(dropna=True)
# Null is not calculated.

xy    1
zs    1
Name: toy, dtype: int64

In [118]:
df1['toy'].value_counts(dropna=False)
# Null is calculated.

xy     1
zs     1
NaN    1
Name: toy, dtype: int64

In [119]:
# NOL NULL FUNCTION: WHERE VALUE IS NOT NULL.
df1.notnull()
# Returns a boolean value where value is not null.

name    toy   born
0  True   True  False
1  True   True   True
2  True  False  False

# Opening and writing in excel file:

For this we need to install 2 modules in terminal as:

Module 1: **xlrd**: pip install xlrd 

This module is needed to read an excel file

Module 2: 

In [120]:
data= pd.read_excel("Book1.xlsx")

In [121]:
data
# By default sheet1 comes into read.

Unnamed: 0.23  Unnamed: 0.22  Unnamed: 0.21  Unnamed: 0.20  Unnamed: 0.19  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   Unnamed: 0.18  Unnamed: 0.17  Unnamed: 0.16  Unnamed: 0.15  Unnamed: 0.14  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   ...  Unnamed: 0.6  Unnamed: 0.5  Unnamed: 0.4  Unnamed: 0.3  Unnamed: 0.2  \
0  ...             0             0             0             0             0   
1  ...             1             1             1             1             1   
2  ...             2             2             2             2             2   
3  ...             3             3             3             3             3   

   Unnamed: 0.1  Unnamed: 0  name   marks  city  
0             0           0     a  123456     p  
1             1           1     b     223     q  
2             2           2     c      34     r  
3             3           3     d     456     s  

[4 rows x 27 columns]

In [122]:
data1=pd.read_excel("Book1.xlsx",sheet_name='Sheet2')
# Here we will get the second sheet.

In [123]:
data1

Unnamed: 0.23  Unnamed: 0.22  Unnamed: 0.21  Unnamed: 0.20  Unnamed: 0.19  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   Unnamed: 0.18  Unnamed: 0.17  Unnamed: 0.16  Unnamed: 0.15  Unnamed: 0.14  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   ...  Unnamed: 0.6  Unnamed: 0.5  Unnamed: 0.4  Unnamed: 0.3  Unnamed: 0.2  \
0  ...             0             0             0             0             0   
1  ...             1             1             1             1             1   
2  ...             2             2             2             2             2   
3  ...             3             3             3             3             3   

   Unnamed: 0.1  Unnamed: 0  name   marks  city  
0             0           0     a  123456     p  
1             1           1     b     223     q  
2             2           2     c      34     r  
3             3           3     d     456     s  

[4 rows x 27 columns]

In [124]:
# Changing the data in the sheet2:
data1.loc[0,'marks']=1234

In [125]:
data1

Unnamed: 0.23  Unnamed: 0.22  Unnamed: 0.21  Unnamed: 0.20  Unnamed: 0.19  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   Unnamed: 0.18  Unnamed: 0.17  Unnamed: 0.16  Unnamed: 0.15  Unnamed: 0.14  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   ...  Unnamed: 0.6  Unnamed: 0.5  Unnamed: 0.4  Unnamed: 0.3  Unnamed: 0.2  \
0  ...             0             0             0             0             0   
1  ...             1             1             1             1             1   
2  ...             2             2             2             2             2   
3  ...             3             3             3             3             3   

   Unnamed: 0.1  Unnamed: 0  name  marks  city  
0             0           0     a   1234     p  
1             1           1     b    223     q  
2             2           2     c     34     r  
3             3           3     d    456     s  

[4 rows x 27 columns]

In [126]:
# Now to wrtie the changed data in the excel file we need:
data1.to_excel("Book1.xlsx",sheet_name='Sheet2')
# Data is changed but sheet 1 is gone.

In [127]:
data2=pd.read_excel("Book1.xlsx",sheet_name='Sheet2')
# Here we will get the second sheet.

In [128]:
data2

Unnamed: 0.24  Unnamed: 0.23  Unnamed: 0.22  Unnamed: 0.21  Unnamed: 0.20  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   Unnamed: 0.19  Unnamed: 0.18  Unnamed: 0.17  Unnamed: 0.16  Unnamed: 0.15  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   ...  Unnamed: 0.6  Unnamed: 0.5  Unnamed: 0.4  Unnamed: 0.3  Unnamed: 0.2  \
0  ...             0             0             0             0             0   
1  ...             1             1             1             1             1   
2  ...             2             2             2             2             2   
3  ...             3             3             3             3             3   

   Unnamed: 0.1  Unnamed: 0  name  marks  city  
0             0           0     a   1234     p  
1             1           1     b    223     q  
2             2           2     c     34     r  
3             3           3     d    456     s  

[4 rows x 28 columns]

In [129]:
# Changing the data in the sheet2:
data2.loc[0,'marks']=123456

In [130]:
data2

Unnamed: 0.24  Unnamed: 0.23  Unnamed: 0.22  Unnamed: 0.21  Unnamed: 0.20  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   Unnamed: 0.19  Unnamed: 0.18  Unnamed: 0.17  Unnamed: 0.16  Unnamed: 0.15  \
0              0              0              0              0              0   
1              1              1              1              1              1   
2              2              2              2              2              2   
3              3              3              3              3              3   

   ...  Unnamed: 0.6  Unnamed: 0.5  Unnamed: 0.4  Unnamed: 0.3  Unnamed: 0.2  \
0  ...             0             0             0             0             0   
1  ...             1             1             1             1             1   
2  ...             2             2             2             2             2   
3  ...             3             3             3             3             3   

   Unnamed: 0.1  Unnamed: 0  name   marks  city  
0             0           0     a  123456     p  
1             1           1     b     223     q  
2             2           2     c      34     r  
3             3           3     d     456     s  

[4 rows x 28 columns]

In [131]:
with pd.ExcelWriter('Book1.xlsx', engine="openpyxl", mode="a" ,if_sheet_exists="overlay") as writer:
    data2.to_excel(writer, sheet_name="Sheet2")

# Now data is updated as per the new sheet. 

HOW TO **MERGE** A DATAFRAME:


To merge a dataframe we need to use a function as .merge()


Mostly we merge the data when 2 dataframe have same column with same values and we need to see the complete relationship of all the columns.

Without right on and left on this will not work:

data_3=pd.merge(data_1,data_2)
MergeError: No common columns to perform merge on. Merge options: left_on=None, right_on=None, left_index=False, right_index=False

**CORRECT SYNTAX:**

data_3=pd.merge(data_1,data_2,right_on='Y',left_on='B')

**Hence it is mendatory to have a common column in 2 data frame to get it merged.**

Here we have changed the data format to seee the merge function here we can see that the function is mapping the data.

That means if **Y column and B column both should have the same data.**

data_6=pd.merge(data_4,data_5,right_on='Y',left_on='B')

data_6

In [132]:
data_1=pd.DataFrame([[1,2,3],[4,5,6],[7,8,9]], columns=['A','B','C'])

In [133]:
data_1

A  B  C
0  1  2  3
1  4  5  6
2  7  8  9

In [134]:
data_2=pd.DataFrame([[11,2,31],[42,5,63],[78,8,95]], columns=['X','Y','Z'])

In [135]:
data_2

X  Y   Z
0  11  2  31
1  42  5  63
2  78  8  95

In [136]:
# Now while merginh we can see that columns B and column Y have same values. 
data_3=pd.merge(data_1,data_2,right_on='Y',left_on='B')

In [137]:
pd.merge?

In [138]:
data_3

A  B  C   X  Y   Z
0  1  2  3  11  2  31
1  4  5  6  42  5  63
2  7  8  9  78  8  95

In [139]:
# Now if we change the data a little then:
data_4=pd.DataFrame([[1,5,3],[4,2,6],[7,8,9]], columns=['A','B','C'])
data_5=pd.DataFrame([[11,2,31],[42,5,63],[78,8,95]], columns=['X','Y','Z'])
data_4

A  B  C
0  1  5  3
1  4  2  6
2  7  8  9

In [140]:
data_5

X  Y   Z
0  11  2  31
1  42  5  63
2  78  8  95

In [141]:
data_6=pd.merge(data_4,data_5,right_on='Y',left_on='B')
data_6

A  B  C   X  Y   Z
0  1  5  3  42  5  63
1  4  2  6  11  2  31
2  7  8  9  78  8  95

**VERY IMPORTANT INFORMATION:**



In [142]:
d1={'Name':['a','b','c','d','e','f','g'],
   'Marks':[99,98,90,78,45,100,87],
   'Gender':['m','f','m','f','m','f','m']}

In [143]:
df1=pd.DataFrame(d1)

In [144]:
df1

Name  Marks Gender
0    a     99      m
1    b     98      f
2    c     90      m
3    d     78      f
4    e     45      m
5    f    100      f
6    g     87      m

### Display Top 3 rows of the dataframe:

In [145]:
df1.head(3)

Name  Marks Gender
0    a     99      m
1    b     98      f
2    c     90      m

### Display last 3 rows of the dataframe

In [146]:
df1.tail(3)

Name  Marks Gender
4    e     45      m
5    f    100      f
6    g     87      m

### Find the shape if the dataframe(Number of Rows and Number of Columns)

In [147]:
df1.shape
# Here shape is not a method it is a attribute of the dataframe.

(7, 3)

In [148]:
print('Number of rows:',df1.shape[0])
print('Number of columns:',df1.shape[1])

Number of rows: 7
Number of columns: 3


### Get information about our dataset like total number of rows and total number of columns datatypes of each columns and memory requirement ###

In [149]:
df1.info()
# 7 rows from 0 to 6
# Name Marks and Gender having 7 entries that is rows 
# Name and Gender are of Object type.
# Marks are of int64 type
# All the columns are having values hence all are not null.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    7 non-null      object
 1   Marks   7 non-null      int64 
 2   Gender  7 non-null      object
dtypes: int64(1), object(2)
memory usage: 296.0+ bytes


### Check NULL values in the Dataset

In [150]:
df1.isnull()
# False which means that all tha values in the table are not null.

Name  Marks  Gender
0  False  False   False
1  False  False   False
2  False  False   False
3  False  False   False
4  False  False   False
5  False  False   False
6  False  False   False

In [151]:
df1.isnull().sum()
# Sum of the total number of null values in the columns.

Name      0
Marks     0
Gender    0
dtype: int64

In [152]:
df1.isnull().sum(axis=1)
# Sum of the total number of null values in the rows.

0    0
1    0
2    0
3    0
4    0
5    0
6    0
dtype: int64

### Get overall statistics about the Dataframe

In [153]:
df1.describe()
# By defualt describe method will tell us the data for int or nummerical data only

Marks
count    7.000000
mean    85.285714
std     19.439773
min     45.000000
25%     82.500000
50%     90.000000
75%     98.500000
max    100.000000

In [154]:
# If we want the data for the complete table we need to use:
df1.describe(include='all')
# freq: Frequency of the top value.
# 25%: values are below 82
# max: value in marks column is 99

Name       Marks Gender
count     7    7.000000      7
unique    7         NaN      2
top       a         NaN      m
freq      1         NaN      4
mean    NaN   85.285714    NaN
std     NaN   19.439773    NaN
min     NaN   45.000000    NaN
25%     NaN   82.500000    NaN
50%     NaN   90.000000    NaN
75%     NaN   98.500000    NaN
max     NaN  100.000000    NaN

### Unique value in Gender column:

In [155]:
df1['Gender'].unique()

array(['m', 'f'], dtype=object)

### Number of unique values in gender column

In [156]:
df1['Gender'].nunique()

2

### Display count of unique values in gender column

In [157]:
df1['Gender'].value_counts()
# m and f are unique values where m comes 4 times and f comes 3 times

m    4
f    3
Name: Gender, dtype: int64

### Find the total number of students in the table having marks from 90 to 100 inclusive of both using between method.

In [158]:
# Without using between method:
df1['Marks']
# complete column of the marks are seen here

0     99
1     98
2     90
3     78
4     45
5    100
6     87
Name: Marks, dtype: int64

In [159]:
# Now by applying 1 condition:
df1['Marks']>=90
# This will return a boolean value.

0     True
1     True
2     True
3    False
4    False
5     True
6    False
Name: Marks, dtype: bool

In [160]:
# Now to get the row where the data is true we need to:
df1[df1['Marks']>=90]   
# Dataframe data always comes under [].

Name  Marks Gender
0    a     99      m
1    b     98      f
2    c     90      m
5    f    100      f

In [161]:
# Now combining 2 condition together then:

# df1['Marks']>=90 & df1['Marks']<=100

# ValueError                                Traceback (most recent call last)
# Cell In [162], line 3
#       1 # Now combining 2 condition together then:
# ----> 3 df1['Marks']>=90 & df1['Marks']<=100

# File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\pandas\core\generic.py:1527, in NDFrame.__nonzero__(self)
#    1525 @final
#    1526 def __nonzero__(self) -> NoReturn:
# -> 1527     raise ValueError(
#    1528         f"The truth value of a {type(self).__name__} is ambiguous. "
#    1529         "Use a.empty, a.bool(), a.item(), a.any() or a.all()."
#    1530     )

# ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().


In [162]:
# For this we need to put 2 or more condition in bracket:

(df1['Marks']>=90) & (df1['Marks']<=100)

0     True
1     True
2     True
3    False
4    False
5     True
6    False
Name: Marks, dtype: bool

In [163]:
df1[(df1['Marks']>=90) & (df1['Marks']<=100)]

# THIS IS HOW WE FIND THE DATA WITHOUT USING BERWEEN.

Name  Marks Gender
0    a     99      m
1    b     98      f
2    c     90      m
5    f    100      f

In [164]:
# To find the sum of the values which are between 90 to 100:

len(df1[(df1['Marks']>=90) & (df1['Marks']<=100)])

# This is the calculation oftotal number of rows which are satisfying the condition.

4

In [165]:
# BY using between now:

df1['Marks'].between(90,100)

# Here both are inclusive that means that resukt will have 90 and 100 also.

0     True
1     True
2     True
3    False
4    False
5     True
6    False
Name: Marks, dtype: bool

In [166]:
# To calculate the sum of the values which are true then:

sum(df1['Marks'].between(90,100))

# Hence we can see that the result is same.

4

In [167]:
df1[df1['Marks'].between(90,100)]

Name  Marks Gender
0    a     99      m
1    b     98      f
2    c     90      m
5    f    100      f

### Find average marks:

In [168]:
df1

Name  Marks Gender
0    a     99      m
1    b     98      f
2    c     90      m
3    d     78      f
4    e     45      m
5    f    100      f
6    g     87      m

In [169]:
df1['Marks']
# Marks column

0     99
1     98
2     90
3     78
4     45
5    100
6     87
Name: Marks, dtype: int64

In [170]:
df1['Marks'].mean()

# This is the average marks in the table.

85.28571428571429

In [171]:
df1['Marks'].min()
# Minimum marks in the table.

45

In [172]:
df1['Marks'].max()
# Maximum marks in the table.

100

### Using Apply Method:

Here this method is used to apply our **user defined method** in the datafrsme.

In [173]:
# Say we need to apply a function in marks column: We need marks divided by 2 as output

def marks(X):
    return(X/2)

# X/2 is normal division where as X//2 is floor division in floor division we donot get any points or float data.

In [174]:
df1['Marks'].apply(marks)

0    49.5
1    49.0
2    45.0
3    39.0
4    22.5
5    50.0
6    43.5
Name: Marks, dtype: float64

In [175]:
# Now we need to add this column in data frame: Column name as half_marks

df1['Half_Marks']=df1['Marks'].apply(marks)

In [176]:
df1

Name  Marks Gender  Half_Marks
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
3    d     78      f        39.0
4    e     45      m        22.5
5    f    100      f        50.0
6    g     87      m        43.5

**By using lambda function:**

In [177]:
df1['Marks'].apply(lambda x:x/2)
# No need to define a function here.

0    49.5
1    49.0
2    45.0
3    39.0
4    22.5
5    50.0
6    43.5
Name: Marks, dtype: float64

**To find the length of strings in name column:**

In [178]:
df1['Name'].apply(len)

# THis shows the length of strings in name columnas all are 1.

0    1
1    1
2    1
3    1
4    1
5    1
6    1
Name: Name, dtype: int64

### MAP FUNCTION: ###

**This function is needed to encode some data into simpler form like say here in gender male to 1 and female to 0**

In [179]:
df1['Gender'].map({'m':1,'f':0})
# Dictionary is needed to be passed here.

0    1
1    0
2    1
3    0
4    1
5    0
6    1
Name: Gender, dtype: int64

In [180]:
df1['Gender_Mapping']=df1['Gender'].map({'m':1,'f':0})

In [181]:
df1

Name  Marks Gender  Half_Marks  Gender_Mapping
0    a     99      m        49.5               1
1    b     98      f        49.0               0
2    c     90      m        45.0               1
3    d     78      f        39.0               0
4    e     45      m        22.5               1
5    f    100      f        50.0               0
6    g     87      m        43.5               1

### Dropping the columns:

In [182]:
df1

Name  Marks Gender  Half_Marks  Gender_Mapping
0    a     99      m        49.5               1
1    b     98      f        49.0               0
2    c     90      m        45.0               1
3    d     78      f        39.0               0
4    e     45      m        22.5               1
5    f    100      f        50.0               0
6    g     87      m        43.5               1

In [183]:
# When dropping only 1 column:

df1.drop('Gender_Mapping',axis=1)

Name  Marks Gender  Half_Marks
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
3    d     78      f        39.0
4    e     45      m        22.5
5    f    100      f        50.0
6    g     87      m        43.5

In [184]:
# When we need to drop multiple columns:

df1.drop(['Gender','Half_Marks'],axis=1)

Name  Marks  Gender_Mapping
0    a     99               1
1    b     98               0
2    c     90               1
3    d     78               0
4    e     45               1
5    f    100               0
6    g     87               1

In [185]:
df1

# Now here we can see that the dataframe has not changed:


Name  Marks Gender  Half_Marks  Gender_Mapping
0    a     99      m        49.5               1
1    b     98      f        49.0               0
2    c     90      m        45.0               1
3    d     78      f        39.0               0
4    e     45      m        22.5               1
5    f    100      f        50.0               0
6    g     87      m        43.5               1

In [186]:
df1=df1.drop(['Gender_Mapping'],axis=1)

In [187]:
df1

# Now changed forever

Name  Marks Gender  Half_Marks
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
3    d     78      f        39.0
4    e     45      m        22.5
5    f    100      f        50.0
6    g     87      m        43.5

In [188]:
df1.loc[0,0]=23.2

In [189]:
df1

Name  Marks Gender  Half_Marks     0
0    a     99      m        49.5  23.2
1    b     98      f        49.0   NaN
2    c     90      m        45.0   NaN
3    d     78      f        39.0   NaN
4    e     45      m        22.5   NaN
5    f    100      f        50.0   NaN
6    g     87      m        43.5   NaN

In [190]:
# Now need to drop this 0 column:

df1.drop([0],axis=1,inplace=True)

In [191]:
df1

# Same column dropped forever.

Name  Marks Gender  Half_Marks
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
3    d     78      f        39.0
4    e     45      m        22.5
5    f    100      f        50.0
6    g     87      m        43.5

### Print name of the columns:

In [192]:
df1.columns

# This is not a method but an attribute of the dataframe.

Index(['Name', 'Marks', 'Gender', 'Half_Marks'], dtype='object')

In [193]:
# Display row index for the dataframe.

df1.index

RangeIndex(start=0, stop=7, step=1)

### Sorting data in the dataframe:

In [194]:
df1

Name  Marks Gender  Half_Marks
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
3    d     78      f        39.0
4    e     45      m        22.5
5    f    100      f        50.0
6    g     87      m        43.5

In [195]:
df1.sort_values(by='Marks')
# Marks by default sorted in assending order.
# Now definition of the function can be checked with shift + tab command.

Name  Marks Gender  Half_Marks
4    e     45      m        22.5
3    d     78      f        39.0
6    g     87      m        43.5
2    c     90      m        45.0
1    b     98      f        49.0
0    a     99      m        49.5
5    f    100      f        50.0

In [196]:
df1.sort_values(by='Marks',ascending=False)
# Hence marks are sorted in decending order.

Name  Marks Gender  Half_Marks
5    f    100      f        50.0
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
6    g     87      m        43.5
3    d     78      f        39.0
4    e     45      m        22.5

In [197]:
# Now sorting multiple columns:
df1.sort_values(by=['Marks','Gender'],ascending=False)
# First desending sortin terms of marks and then gender.

Name  Marks Gender  Half_Marks
5    f    100      f        50.0
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
6    g     87      m        43.5
3    d     78      f        39.0
4    e     45      m        22.5

### Display name and marks of female candidate only.

In [198]:
df1

Name  Marks Gender  Half_Marks
0    a     99      m        49.5
1    b     98      f        49.0
2    c     90      m        45.0
3    d     78      f        39.0
4    e     45      m        22.5
5    f    100      f        50.0
6    g     87      m        43.5

In [199]:
df1['Gender']=='f'

0    False
1     True
2    False
3     True
4    False
5     True
6    False
Name: Gender, dtype: bool

In [200]:
df1[df1['Gender']=='f']

Name  Marks Gender  Half_Marks
1    b     98      f        49.0
3    d     78      f        39.0
5    f    100      f        50.0

In [201]:
# Just need name and marks:
df1[df1['Gender']=='f']['Name']

1    b
3    d
5    f
Name: Name, dtype: object

In [202]:
df1[df1['Gender']=='f'][['Name','Marks']]
# 2-D list.

Name  Marks
1    b     98
3    d     78
5    f    100

In [203]:
# Name marks and gender:
df1[df1['Gender']=='f'][['Name','Marks','Gender']]

Name  Marks Gender
1    b     98      f
3    d     78      f
5    f    100      f

#### Using isin method for the same process:

In [204]:
df1['Gender'].isin(['f'])
# Boolean series generated need to put in dataframe will get the index complete for the condition.

0    False
1     True
2    False
3     True
4    False
5     True
6    False
Name: Gender, dtype: bool

In [205]:
df1[df1['Gender'].isin(['f'])]

Name  Marks Gender  Half_Marks
1    b     98      f        49.0
3    d     78      f        39.0
5    f    100      f        50.0

In [206]:
df1[df1['Gender'].isin(['f'])][['Name','Marks']]

Name  Marks
1    b     98
3    d     78
5    f    100